### ISSUE TITLE
Logging system tests are hanging on CircleCI

### ISSUE THREAD

**Issue by dhermes** [GoogleCloudPlatform/google-cloud-python#3237]
Four builds all hung:

- https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1413 (after `.........` == 9)
- https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1415 (after `.............` == 13)
- https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1418 (after `...........` == 11)
- https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1426 (after `......` == 6)

I have run these tests 8+ times locally since the original failure (and using the same Cloud Console project) and they run without issue.

@waprin Any ideas?

**Comment by dhermes** [comment]
Note this is "blocking" the release of `error_reporting == 0.23.3` /cc @DizzeePascall

**Comment by dhermes** [comment]
Running `py.test tests/system.py -vv` (under the `logging` directory) the (deterministic) order of test cases is:

```
01 tests/system.py::TestLogging::test_create_metric PASSED
02 tests/system.py::TestLogging::test_create_sink_bigquery_dataset PASSED
03 tests/system.py::TestLogging::test_create_sink_pubsub_topic PASSED
04 tests/system.py::TestLogging::test_create_sink_storage_bucket PASSED
05 tests/system.py::TestLogging::test_list_metrics PASSED
06 tests/system.py::TestLogging::test_list_sinks PASSED
07 tests/system.py::TestLogging::test_log_handler_async PASSED
08 tests/system.py::TestLogging::test_log_handler_sync PASSED
09 tests/system.py::TestLogging::test_log_root_handler PASSED
10 tests/system.py::TestLogging::test_log_struct PASSED
11 tests/system.py::TestLogging::test_log_struct_w_metadata PASSED
12 tests/system.py::TestLogging::test_log_text PASSED
13 tests/system.py::TestLogging::test_log_text_w_metadata PASSED
14 tests/system.py::TestLogging::test_log_text_with_timestamp PASSED
15 tests/system.py::TestLogging::test_reload_metric PASSED
16 tests/system.py::TestLogging::test_reload_sink PASSED
17 tests/system.py::TestLogging::test_update_metric PASSED
18 tests/system.py::TestLogging::test_update_sink PASSED
```

(with numbers added by me).

**Comment by waprin** [comment]
I'm a bit confused as to under what circumstances these tests are failing? Is it with those error reporting changes or is it just sporadically failing everywhere (sorry Circle interface isn't obviously answering this question to me).

 `test_log_root_handler(self)` is the most complex test and worth commenting out to see if it fixes anything.

**Comment by dhermes** [comment]
The `logging` environment is completely isolated from the `error_reporting` changes, so they are unrelated. (Though `logging` is a dependency of `error_reporting`, so we want to make sure `logging` is good before doing a downstream release.)

Every CI run since the first failure has failed, i.e. it doesn't seem sporadic. To replicate we'd want to use the same env. vars. locally?

**Comment by dhermes** [comment]
@waprin We've now got a fourth instance.

**Comment by lukesneeringer** [comment]
Adding an `issue-3237` branch with `-vvv` so that we can see it more easily.

**Comment by dhermes** [comment]
![55452314](https://cloud.githubusercontent.com/assets/520669/24512861/8d1ed37c-1524-11e7-9066-60e466c1165b.jpg)

**Comment by lukesneeringer** [comment]
Closing, but merging #3244 to help us watch.

**Comment by dhermes** [comment]
This issue has [come back][1] (and [here][2] and [here][3])

[1]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1780
[2]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1774
[3]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1753

**Comment by dhermes** [comment]
@waprin @jonparrott This is somewhat concerning. Do you guys have any ideas what could be causing this?

**Comment by jonparrott** [comment]
I really have no idea - what's our gRPC deadline?

**Comment by dhermes** [comment]
Whatever the default is? It **consistently** hangs though. A "bad" deadline wouldn't happen twice a day?

**Comment by jonparrott** [comment]
If the RPC itself is hanging and our deadline is ~10 minutes, I can see how
this occurs.

Has anyone been able to reproduce locally?

**Comment by waprin** [comment]
While I'd like to imagine my code is bug-free, and I have no specific idea of where a bug might be, the piece of code I'd be by far most suspicious of causing a deadlock is the async `background_thread.py` logging handler transport since it starts a new thread, and both the main thread and background thread acquire locks. I had to fix a few deadlocks when I first wrote it.

I would suggest commenting out `test_log_handler_async` and `test_log_root_handler` and see if these deadlocks still happen. If not, we've narrowed it down quite a bit. We could also switch the default transport to the synchronous option until bugs are ironed out.

I think the log handlers are important and a backgrounded writer is important but it is always tricky to get multi-threaded code right, probably should have marked those in some sort of `contrib` package in first place. We had discussed moving it out of this repo altogether but I think standard logging handlers prob do belong in here, in either case it seems like @duggelz and @liyanhui1228 are building on top of it and taking ownership of some of these sort of features, so maybe they can take a look and help iron out any kinks?

Again, this is pure speculation because I don't actually have any reason to think there are bugs in that code other than its an easy piece of code to have subtle hard-to-find, hard-to-repro bugs in.

**Comment by jonparrott** [comment]
What locking is even needed by the background transport? It's just a queue,
right?

**Comment by waprin** [comment]
Pretty much but `Queue` isn't quite what we want since you need to commit all the items at once rather than pick them off one a time. You also need some locks to gracefully shutdown regardless.

**Comment by jonparrott** [comment]
Looks like the background transport could use some love, I'm not sure the conditions are the best way to accomplish what we want.

Even still, it seems we're still kind of guessing at what's going on. Again, has anyone reproduced this locally?

**Comment by duggelz** [comment]
It might be helpful to add 'faulthandler.enable()' or '.register()' in our system test harness to get a multi-thread traceback when the test is killed.  We'll have to figure out what signals CircleCI uses to kill jobs (and hopefully it's something reasonable).

**Comment by dhermes** [comment]
@waprin Regarding your [last comment][1], I have annotated my [previous comment][2] to show exactly which methods are / were hanging.

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3237#issuecomment-300899965
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/issues/3237#issuecomment-300285940

**Comment by dhermes** [comment]
I'm going to take a stab here and add some print statements (I'm worried about using Python `logging` since this code hooks into it).

**Comment by duggelz** [comment]
[faulthandler](https://docs.python.org/3/library/faulthandler.html) prints to standard error directly, so it should be safe.  It's only available in Python 3 though (there is a backport to 2.7).

**Comment by duggelz** [comment]
I'm able to reproduce some logging test errors in the internal Jenkins builds for the [python-runtime](https://github.com/GoogleCloudPlatform/python-runtime) repo, but nothing when running locally.

**Comment by duggelz** [comment]
Just FYI, `faulthandler` does have some deadlock problems after all: https://bugs.python.org/issue30165

**Comment by dhermes** [comment]
Just got a [build][1] that seems to shed some light on the issue. The issue seems to be in the tear down:

```python
    def tearDown(self):
        retry = RetryErrors(NotFound, max_tries=10)
        for doomed in self.to_delete:
            retry(doomed.delete)()
        logging.getLogger().handlers = self._handlers_cache[:]
```

Using `max_tries=10` means we can have a total sleep of

```
1 + 2 + ... + 256 + 512 = 1024 - 1
```

This **exceeds** the CircleCI threshold of 600 seconds (10 minutes) without any output.

[1]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/1868

**Comment by dhermes** [comment]
@jonparrott Do you feel OK if I change this back to closed?

**Comment by jonparrott** [comment]
Close it!

**Comment by dhermes** [comment]
Blerg. I just saw [another one][1]

[1]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/2072

**Comment by dhermes** [comment]
And [another one][1]

[1]: https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/2091

**Comment by dhermes** [comment]
@lukesneeringer This is still a **huge** problem.

**Comment by jonparrott** [comment]
This is now outright failures. We should get this fixed.

**Comment by jonparrott** [comment]
Whoever takes this on should wait until after #4759 is merged.

**Comment by jonparrott** [comment]
@chemelnucfin can I task you with taking care of the low-hanging fruit here? (there's some tests that refer to old bigquery names, so should be trivial to fix).

In terms of hanging, see if you can still make it happen after #4759 & I'll see what I can do to get it fixed.

**Comment by chemelnucfin** [comment]
I saw the merge, working on the bigquery ones.

**Comment by chemelnucfin** [comment]
Possibly fixed by #4768.  Will reopen if that's not the case.
